### DISCUSSION TITLE
Zod: allow `undefined` as default value for a required field

### DISCUSSION THREAD

**Discussion by knoefel** [colinhacks/zod#1198] [Q&A]
I have a simple form to "add/edit" an entity. My (simplified) `zod` schema looks like this:

```typescript
const schema = z.object({
  host: z.string(),
  port: z.number()
});
```

The default values for the "add" use case look like this:

```typescript
const defaultValues = {
  host: "Host"
};
```

**These are my requirements for the `port` field:**
- the field is required
- the field value is a number
- upon the "add" use case, it should be possible to provide an empty field to the user (so it would fail if the user hits submit before entering a value)
- it should be possible to set default values for other fields in the schema, without setting a value for the `port` field

So for the "add" use case most of the form fields should be empty (except the `host` field), only when i edit i want to set the current values for that entity as the default values.

The problem is, that `zod` won't allow this, as i must set a default value for the `port` field as it is a required field. From an UX perspective this isn't optimal, as i want to leave the `port` field empty and force the user to enter a value.

The only way i've found so far is to set the `port` field to `0`. But for the reason above this is not a good solution.

I wonder if there are other solutions to this problem? It seems like this shouldn't be rare case, as it is usually a requirement to pre-fill some fields, while other's must explicitly filled out by the user.

**Comment by ecyrbe** [comment]
Just distinct form completion from form validation.

**Reply by AnderssonPeter** [reply]
While this might be the correct solution, it's not totally clear what this means, a simple example would go a long way!

**Reply by Karsenh** [reply]
Literally, what does "just distinct form completion from form validation" even mean lol?

**Comment by bloodykheeng** [comment]
try to use refine

```
const formSchema = z.object({
  name: z.string().min(3, "Name must be at least 3 characters").refine((val) => !!val, {
    message: "Name is required",
  }),

  cso: z.object({ id: z.number(), name: z.string() }).nullable().refine((val) => !!val === false, {
    message: "CSO is required",
  }),
})


// ✅ Default Form Values - Make sure types match exactly
const defaultValues: FormData = {
  name: "",
  cso: null,
}
```

**Comment by bloodykheeng** [comment]
sorry seems refine doesnt work well use superrefine

```

const formSchema = z.object({ 
  name: z.string().min(3, "Name must be at least 3 characters").refine((val) => !!val, { 
    message: "Name is required", 
  }), 
 
  cso: z.object({ id: z.number(), name: z.string() }) 
    .nullable()
    .superRefine((val, ctx) => {
      if (!val) {
        ctx.addIssue({
          code: z.ZodIssueCode.custom,
          message: "CSO is required",
        });
        return z.NEVER;
      }
      return val;
    })
});
```

**Comment by PizzaDependency** [comment]
Im surprised this is from 2022 and still not proper solution just a workarounds
This is what i did:
```js
const schema = z.object({
  host: z.string(),
  port: z.number()
});
```
```js
const defaultValues = {
  host: "Host",
  port: "" as unknown as number
};
```
With this you wont need to write any extra code inside `refine` or `superRefine`, and type wont be messed up accepting strings.
